File: BEditor/BEditor.CLI/Point3.cpp
#include "pch.h"

using namespace System;

using namespace BEditor::CLI::Media;

inline Point3::Point3(int x, int y, int z) {
	X = x;
	Y = y;
	Z = z;
}
inline Point3::Point3(float x, float y, float z) {
	X = x;
	Y = y;
	Z = z;
}
inline Point3::Point3(double x, double y, double z) {
	X = (float)x;
	Y = (float)y;
	Z = (float)z;
}

inline Point3 Point3::Add(Point3 point, Size size) {
	return Point3(
		point.X + size.Width,
		point.Y + size.Height,
		point.Z);
}
inline Point3 Point3::Add(Point3 point1, Point3 point2) {
	return Point3(
		point1.X + point2.X,
		point1.Y + point2.Y,
		point1.Z + point2.Z);
}
inline Point3 Point3::Subtract(Point3 point, Size size) {
	return Point3(
		point.X - size.Width,
		point.Y - size.Height,
		point.Z);
}
inline Point3 Point3::Subtract(Point3 point1, Point3 point2) {
	return Point3(
		point1.X - point2.X,
		point1.Y - point2.Y,
		point1.Z - point2.Z);
}
inline bool Point3::Equals(Object^ obj) {
	return obj->GetType() == Point3::typeid && Equals((Point3)obj);
}
inline bool Point3::Equals(Point3 point) {
	return
		X == point.X &&
		Y == point.Y &&
		Z == point.Z;
}
inline int Point3::GetHashCode() {
	return HashCode::Combine(X, Y, Z);
}
inline String^ Point3::ToString() {
	return String::Format("(X:{0} Y:{1} Z{2})", X, Y, Z);
}

inline Point3 Point3::operator+(Point3 point1, Point3 point2) {
	return Add(point1, point2);
}
inline Point3 Point3::operator+(Point3 point, Size size) {
	return Add(point, size);
}
inline Point3 Point3::operator-(Point3 point1, Point3 point2) {
	return Subtract(point1, point2);
}
inline Point3 Point3::operator-(Point3 point, Size size) {
	return Subtract(point, size);
}
inline bool Point3::operator==(Point3 left, Point3 right) {
	return left.Equals(right);
}
inline bool Point3::operator!=(Point3 left, Point3 right) {
	return !left.Equals(right);
}

File: src/BEditor.Drawing/Point3.cs
using System;
using System.Runtime.Serialization;

namespace BEditor.Drawing
{
    [Serializable]
    public readonly struct Point3 : IEquatable<Point3>, ISerializable
    {
        public static readonly Point3 Empty;

        public Point3(int x, int y, int z)
        {
            X = x;
            Y = y;
            Z = z;
        }
        private Point3(SerializationInfo info, StreamingContext context)
        {
            X = info.GetInt32(nameof(X));
            Y = info.GetInt32(nameof(Y));
            Z = info.GetInt32(nameof(Z));
        }

        public int X { get; }
        public int Y { get; }
        public int Z { get; }

        public static Point3 Add(Point3 point1, Point3 point2)
            => new(point1.X + point2.X, point1.Y + point2.Y, point1.Z + point2.Z);
        public static Point3 Subtract(Point3 point1, Point3 point2)
            => new(point1.X - point2.X, point1.Y - point2.Y, point1.Z - point2.Z);
        public override bool Equals(object? obj)
            => obj is Point3 point && Equals(point);
        public bool Equals(Point3 other)
            => X == other.X && Y == other.Y && Z == other.Z;
        public override int GetHashCode()
            => HashCode.Combine(X, Y, Z);
        public void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            info.AddValue(nameof(X), X);
            info.AddValue(nameof(Y), Y);
            info.AddValue(nameof(Z), Z);
        }

        public static Point3 operator +(Point3 point1, Point3 point2)
            => Add(point1, point2);
        public static Point3 operator -(Point3 point1, Point3 point2)
            => Subtract(point1, point2);
        public static Point3 operator *(Point3 left, Point3 right)
            => new(left.X * right.X, left.Y * right.Y, left.Z * right.Z);
        public static Point3 operator /(Point3 left, Point3 right)
            => new(left.X / right.X, left.Y / right.Y, left.Z / right.Z);
        public static Point3 operator *(Point3 left, int right)
            => new(left.X * right, left.Y * right, left.Z * right);
        public static Point3 operator /(Point3 left, int right)
            => new(left.X / right, left.Y / right, left.Z / right);
        public static bool operator ==(Point3 left, Point3 right)
            => left.Equals(right);
        public static bool operator !=(Point3 left, Point3 right)
            => !(left == right);
    }
}
